### PR TITLE
[BE][Attention] Code de-dup

### PR DESCRIPTION
**Pull request by malfet**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #139791
* #139763
* __->__ #139784
* #139788

The only difference between `convert_boolean_attn_mask_cudnn` and `convert_boolean_attn_mask` is the value we initialize boolean tensor to
Reduce duplication by introducing `convert_boolean_attn_mask_` that takes `neg_inf` value and make abovementioned implementations are trivial oneline call
Also, as suggested by @Skylion007, replace `at::where(foo->logical_not, -inf, 0)` with `at::where(*foo, 0, -inf)`

### CODE REVIEW

**Review by Skylion007 — COMMENTED** [review]
optimization nit

**Review by Skylion007 — APPROVED** [review]

File: aten/src/ATen/native/transformers/attention.cpp [hunk 1]
```diff
@@ -524,35 +524,29 @@ inline void validate_sdpa_input(
...
   // Pass through
-  if(!attn_mask.has_value()){
+  if (!attn_mask.has_value()) {
     return std::nullopt;
   }
   // Convert boolean mask to additive mask; need to invert mask to indicate what
   // to mask *out*.
   if (attn_mask->dtype() == at::kBool) {
-    return at::where(attn_mask->logical_not(), -std::numeric_limits<double>::infinity(), at::scalar_tensor(0.0, at::TensorOptions().dtype(dtype).device(attn_mask->device())));
+    return at::where(*attn_mask, 0.0, at::scalar_tensor(neg_inf, at::TensorOptions().dtype(dtype).device(attn_mask->device())));
```
Comment (drisspg) [inline aten/src/ATen/native/transformers/attention.cpp:535]: we should probably also have 0.0 be a scalar tensor right? I am not sure why this didnt also need to be updated

**Review by drisspg — APPROVED** [review]

File: aten/src/ATen/native/transformers/attention.cpp (hunk 1 above)
Comment (Skylion007) [reply aten/src/ATen/native/transformers/attention.cpp:535]: Lol, I made this edit on the other PR in the stack.

Comment (malfet) [reply aten/src/ATen/native/transformers/attention.cpp:535]: I suspect only one scalar_tensor is needed to preserve out dtype, other can be inferred from the first.

**Pull request closed without merging**